packages/api/src/controllers/access-control.test.ts

```diff
@@ -0,0 +1,175 @@
+      client.jwtAuth = nonAdminToken;
+      let created: SigningKeyResponsePayload = await client
+        .post("/access-control/signing-key")
+        .then((res) => res.json());
+      let res = await client.get(`/access-control/signing-key/${created.id}`);
+      signingKey = await res.json();
+      const res2 = await client.post("/stream", {
+        name: "test",
+        playbackPolicy: {
+          type: "signed",
+        },
+      });
+      const stream = await res2.json();
+      gatedPlaybackId = stream.playbackId;
+      const res3 = await client.post("/stream", {
+        name: "test",
+        playbackPolicy: {
+          type: "public",
+        },
+      });
+      const publicStream = await res3.json();
+      publicPlaybackId = publicStream.playbackId;
+      let id = uuid();
+      gatedAsset = await db.asset.create({
+        id,
+        name: "test-storage",
+        createdAt: Date.now(),
+        playbackId: await generateUniquePlaybackId(id),
+        playbackPolicy: {
+          type: "signed",
+        },
+        status: {
+          phase: "ready",
+          updatedAt: Date.now(),
+        },
+        userId: nonAdminUser.id,
+      });
+    });
+
+    it("should create a gate stream and allow playback with given public key", async () => {
+      client.jwtAuth = adminToken;
+      const res2 = await client.post("/access-control/gate", {
+        stream: `video+${gatedPlaybackId}`,
+        pub: signingKey.publicKey,
+        type: "jwt",
+      });
+
+      expect(res2.status).toBe(204);
+    });
+
+    it("should not allow playback on not existing streams or public keys", async () => {
+      const res = await client.post("/access-control/gate", {
+        stream: `video+0000000000`,
+        pub: signingKey.publicKey,
+        type: "jwt",
+      });
+      expect(res.status).toBe(403);
+      const res2 = await client.post("/access-control/gate", {
+        stream: `video+${gatedPlaybackId}`,
+        pub: "00000000000000",
+        type: "jwt",
+      });
+      expect(res2.status).toBe(403);
+    });
+
+    it("should not allow playback if stream is gated an pub is missing", async () => {
+      const res2 = await client.post("/access-control/gate", {
+        stream: `video+${gatedPlaybackId}`,
+        type: "jwt",
+      });
+      expect(res2.status).toBe(403);
+    });
+
+    it("should not allow playback if stream and public key does not share the same owner", async () => {
+      const otherStream = await client.post("/stream", {
+        name: "test",
+        playbackPolicy: {
+          type: "signed",
+        },
+      });
+      const otherStreamPlaybackId = (await otherStream.json()).playbackId;
+      const res = await client.post("/access-control/gate", {
+        stream: `video+${otherStreamPlaybackId}`,
+        pub: signingKey.publicKey,
+        type: "jwt",
+      });
+      expect(res.status).toBe(403);
+    });
+
```

Comment:
Can you add some additional test cases?
 - Does not allow playback if signing key is disabled
 - Does not allow playback if signing key is deleted (or doesn't even exist)
 - Does not allow playback if the `playbackPolicy` is something we don't understand (this is necessary for bumping the version for example. maybe we find something we don't know and we should fail-closed, not open!)

And maybe we should consider implementing support for not allowing playback if the stream or the user are suspended. WDYT?

Reply:
Sounds good, added!

about not allowing playback if the stream or the user are suspended, wouldn't the stream be interrupted anyway so impossible to playback anyway?

Reply:
This is another case of layered security [1]. Just because there is _theoretically_ another system that is going to do the blocking, it doesn't mean it's detrimental to add more redundant layers here that would make it more effective. Like that idea of listing 2 elements when only 1 is expected and ensuring that only 1 is returned. Just because we expect something to behave in a certain way does not mean that there aren't any bugs or incidents :)

In fact, the current system for suspending streams and users is very much fallible, and the API for it is broken **right now**. We only have an idea of how this is gonna be better in the future, with events about the streams sent to catalyst etc, but it is only an idea right now. 

By adding these checks here, we will "immediately" (after the `/gate` cache expires) block access to all viewers watching the stream. Even if the stream suspension fails, which it very often does, the viewers will still get blocked and then the stream become useless for the pirates.

[1] The original term is for like more hardcore security, but I think it does apply here: https://blog.totalprosource.com/what-is-layered-security-how-does-it-defend-your-network

Reply:
Cool! Thanks for the write up and the info, it makes sense, although eventually, if the user is not actually suspended on Mist and he is still able to stream, he would just need to make the stream public again to keep it streaming and playback normally.

I added the check and a test for the suspended users!

Reply:
The playback should be blocked for suspended users/streams even if the given content is public. Isn't this new API called even for public assets/streams?